Test/Unit/Connection.php

```diff
@@ -0,0 +1,1333 @@
+                ->variable($calledA)
+                    ->isNull()
+                ->boolean($calledB)
+                    ->isTrue()
+                ->variable($calledC)
+                    ->isNull();
+    }
+
+    public function case_run_pong_opcode()
+    {
+        $this
+            ->given(
+                $socket     = new \Mock\Hoa\Socket\Client('tcp://*:1234'),
+                $connection = new SUT($socket),
+                $protocol   = new \Mock\Hoa\Websocket\Protocol\Rfc6455($socket),
+                $this->mockGenerator->orphanize('__construct'),
+                $node = new \Mock\Hoa\Websocket\Node(),
+                $node->setProtocolImplementation($protocol),
+
+                $this->calling($node)->getHandshake  = SUCCEED,
+                $this->calling($protocol)->readFrame = [
+                    'fin'     => 0x1,
+                    'rsv1'    => 0x0,
+                    'rsv2'    => 0x0,
+                    'rsv3'    => 0x0,
+                    'opcode'  => SUT::OPCODE_PONG,
+                    'mask'    => 0x1,
+                    'length'  => 6,
+                    'message' => 'foobar'
+                ],
+                $this->calling($protocol)->writeFrame = function ($_message, $_opcode, $_mask) use (&$called) {
+                    $called = true;
+
+                    return;
+                }
+            )
+            ->when($result = $this->invoke($connection)->_run($node))
+            ->then
+                ->variable($result)
+                    ->isNull()
+                ->variable($called)
+                    ->isNull();
+    }
+
+    public function case_run_pong_opcode_not_fin()
+    {
+        $self = $this;
+
+        $this
+            ->given(
+                $socket     = new \Mock\Hoa\Socket\Client('tcp://*:1234'),
+                $connection = new SUT($socket),
+                $protocol   = new \Mock\Hoa\Websocket\Protocol\Rfc6455($socket),
+                $this->mockGenerator->orphanize('__construct'),
+                $node = new \Mock\Hoa\Websocket\Node(),
+                $node->setProtocolImplementation($protocol),
+
+                $this->calling($node)->getHandshake  = SUCCEED,
+                $this->calling($protocol)->readFrame = [
+                    'fin'     => 0x0,
+                    'rsv1'    => 0x0,
+                    'rsv2'    => 0x0,
+                    'rsv3'    => 0x0,
+                    'opcode'  => SUT::OPCODE_PONG,
+                    'mask'    => 0x1,
+                    'length'  => 6,
+                    'message' => 'foobar'
+                ],
+                $this->calling($protocol)->writeFrame = function ($_message, $_opcode, $_mask) use (&$calledA) {
+                    $calledA = true;
+
+                    return;
+                },
+                $this->calling($connection)->close = function ($_code, $_reason) use (&$calledB, $self) {
+                    $calledB = true;
+
+                    $self
```

Comment:
using mock assertions (`->mock()->call()->withArguments()->exactly()`) you can avoid the ˋ$self` and manual assertions in stubs